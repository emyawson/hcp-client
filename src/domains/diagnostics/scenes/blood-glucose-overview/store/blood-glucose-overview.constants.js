import { translate } from 'src/i18n';

export const VARIABILITY_THRESHOLDS = {
  MEDIUM: 33,
  HIGH: 50,
};

export const TRAFFIC_LIGHT_COLORS = {
  GRAY: 'GRAY',
  GREEN: 'GREEN',
  RED: 'RED',
  YELLOW: 'YELLOW',
};

export const NO_MEASUREMENTS_INTERVAL_LABEL = '-';

export const TRAFFIC_LIGHT_LABELS = {
  HYPO: translate('bloodGlucoseOverview.hypo'),
  HYPER: translate('bloodGlucoseOverview.hyper'),
  IN_RANGE: translate('bloodGlucoseOverview.inRange'),
  LOW: translate('bloodGlucoseOverview.low'),
  MEDIUM: translate('bloodGlucoseOverview.medium'),
  HIGH: translate('bloodGlucoseOverview.high'),
  INSUFFICIENT_INFO: translate('bloodGlucoseOverview.na'),
  NO_INFO: NO_MEASUREMENTS_INTERVAL_LABEL,
};

export const MINIMUM_TESTS_FOR_INTERVAL = 28;
export const MINIMUM_AVERAGE_TESTS_FOR_DAY = 2;

export const NUMBER_BLOOD_GLUCOSE_OVERVIEW_COLUMNS = 6;

export const HYPO_RISK_THRESHOLDS = {
  MEDIUM: 1.1,
  HIGH: 2.5,
};
