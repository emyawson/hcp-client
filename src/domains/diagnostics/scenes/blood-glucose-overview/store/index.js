export {
  bloodGlucoseOverviewConnector,
  calculateAndFormatVariability,
  selectBloodGlucoseOverviewData,
  selectDayInHalfTimeIntervals,
  selectFormattedDateRanges,
  selectHypoglycaemia,
  selectHypoRisk,
  selectIsThereNextInterval,
  selectIsTherePrevInterval,
  selectMeanBloodGlucoseStatistics,
  selectMeanBloodGlucoseStatus,
  selectTestsPerDay,
  selectVerifiedGroupMeasurements,
} from './blood-glucose-overview.selector';
export {
  HYPO_RISK_MAX,
  HYPO_RISK_MIN,
  HYPO_RISK_THRESHOLDS,
  MIN_BG,
  MINIMUM_AVERAGE_TESTS_FOR_DAY,
  MINIMUM_TESTS_FOR_INTERVAL,
  NO_MEASUREMENTS_INTERVAL_LABEL,
  NUMBER_BLOOD_GLUCOSE_OVERVIEW_COLUMNS,
  TRAFFIC_LIGHT_COLORS,
  TRAFFIC_LIGHT_LABELS,
  VARIABILITY_THRESHOLDS,
} from './blood-glucose-overview.constants';
export {
  calculateAvgTestsPerDay,
  calculateBloodGlucoseIndex,
  calculateBloodGlucoseStartDate,
  calculateEndDateAfterPagination,
  calculateForVariability,
  calculateHighBloodGlucoseIndex,
  calculateHypoRisk,
  calculateHypoRiskColor,
  calculateHypoRiskForIntervals,
  calculateHypoRiskLabel,
  calculateIntervalEmptyDays,
  calculateLowBloodGlucoseIndex,
  calculateMeanBloodGlucose,
  calculateMeanColor,
  calculateMeanLabel,
  calculateVariabilityStatus,
  createCalculateForHypoglycaemia,
  determineHasReliableInfo,
  formatTestsPerDayAvg,
  groupByTimeInterval,
  hasReliableInfo,
  isFirstInterval,
  isHyperRed,
  isHyperYellow,
  isHypoRed,
  isHypoRiskHigh,
  isHypoRiskLow,
  isHypoRiskMedium,
  isHypoYellow,
  isInRange,
  isLastInterval,
  roundToNDecimalPlaces,
  toHypoglycaemiaCount,
  toHypoRiskStatus,
  toLabelInterval,
  toMeanBloodGlucoseValue,
  toTimeIntervals,
} from './blood-glucose-overview.utils';
