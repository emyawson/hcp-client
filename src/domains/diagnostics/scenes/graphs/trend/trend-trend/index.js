export { TrendTrendContainer } from './trend-trend.container';
