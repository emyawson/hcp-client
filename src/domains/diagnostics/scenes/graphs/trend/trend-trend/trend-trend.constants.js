export const GRAPH_Y_MAX = 400;
export const GRAPH_Y_MIN = 0;
