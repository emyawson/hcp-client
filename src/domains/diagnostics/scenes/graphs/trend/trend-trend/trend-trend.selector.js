import { createSelector, createStructuredSelector } from 'reselect';
import { keys } from 'ramda';

import { toGMTZone } from 'src/domains/diagnostics/utils';
import {
  selectGraphDetails,
  selectShowGridLines,
  selectTargetRange,
  selectThreshold,
  selectVerticalTicks,
  selectVerticalLabel,
  selectVerticalAxesCeiling,
} from 'src/domains/diagnostics/scenes/graphs/graph.selector';
import {
  GRAPH_Y_MIN,
  GRAPH_Y_MAX,
} from 'src/domains/diagnostics/scenes/graphs/graph.constants';
import { EMPTY_VALUE_PLACEHOLDER } from 'src/domains/diagnostics/store/constants';
import {
  selectPatientStartDate,
  selectPatientEndDate,
  selectGlucoseMeasurementsInDateSliderRange,
  selectGraphLoading,
} from 'src/domains/diagnostics/store/selectors';

import {
  selectHorizontalDayTicks,
  selectHorizontalMonthYearTicks,
} from '../trend.selector';
import {
  fromDateString,
  sortAndCalculateDailyStats,
  calculateDifferenceInDays,
} from '../trend.util';

const normalizeStats = (measurements, startDate, endDate, ceil, floor) => {
  const start = startDate.startOf('day');
  const end = endDate.endOf('day');
  const range = calculateDifferenceInDays(start, end);
  const statsByDay = sortAndCalculateDailyStats(measurements);

  return keys(statsByDay).map(key => {
    const date = toGMTZone(fromDateString(key));
    const difference = calculateDifferenceInDays(start, date);
    const deviation =
      statsByDay[key].stdDev === EMPTY_VALUE_PLACEHOLDER
        ? 0
        : statsByDay[key].stdDev / ceil;

    return {
      max: statsByDay[key].max / ceil,
      min: statsByDay[key].min / ceil,
      deviation,
      x: difference / range + 1 / (range * 2),
      y: statsByDay[key].mean / ceil,
      data: statsByDay[key],
    };
  });
};

const normalizeGraphData = (
  measurements,
  startDate,
  endDate,
  graphYMax = GRAPH_Y_MAX,
) => normalizeStats(measurements, startDate, endDate, graphYMax, GRAPH_Y_MIN);

export const selectGraphData = createSelector(
  selectGlucoseMeasurementsInDateSliderRange,
  selectPatientStartDate,
  selectPatientEndDate,
  selectVerticalAxesCeiling,
  normalizeGraphData,
);

export const trendTrendConnector = createStructuredSelector({
  measurements: selectGlucoseMeasurementsInDateSliderRange,
  graphData: selectGraphData,
  verticalLabel: selectVerticalLabel,
  verticalTicks: selectVerticalTicks,
  horizontalDayTicks: selectHorizontalDayTicks,
  horizontalMonthYearTicks: selectHorizontalMonthYearTicks,
  targetRange: selectTargetRange,
  threshold: selectThreshold,
  showGridLines: selectShowGridLines,
  graphDetails: selectGraphDetails,
  graphYMax: selectVerticalAxesCeiling,
  isLoading: selectGraphLoading,
});
