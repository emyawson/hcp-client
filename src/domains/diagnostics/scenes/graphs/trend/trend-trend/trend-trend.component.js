import React from 'react';
import { isEmpty } from 'ramda';

import { GraphDetail, ToolTip } from 'src/domains/diagnostics/components';
import {
  ResizeGraphWrapper,
  TrendToolTip,
} from 'src/domains/diagnostics/components';
import { Graph } from 'src/domains/diagnostics/lib';
import { RenderIf } from 'src/domains/diagnostics/utils';
import { withToolTip } from 'src/domains/diagnostics/utils/with-tool-tip';
import { filterVerticalTicksForViewAndVisibilityTolerance } from 'src/domains/diagnostics/utils/graphs.util';
import { withGraphLoader } from 'src/domains/diagnostics/utils';
import { EMPTY_VALUE_PLACEHOLDER } from 'src/domains/diagnostics/store/constants';

import { TrendAxesPlot, TrendGraphPlot } from '../components';
import { X_AXIS_WIDTH } from '../trend.constants';
import { TrendPlotWrapperDiv } from '../trend.style';
import { getToolTipValueColor } from '../../graph.util';
import { GraphDetailContainer } from '../../graph.style';
import { Y_AXIS_TICK_VISIBILITY_TOLERANCE } from '../../graph.constants';

const TrendPlotWrapperDivWithLoader = withGraphLoader(TrendPlotWrapperDiv);

export const TrendTrend = withToolTip(
  ({
    measurements,
    graphData,
    graphDetails,
    targetRange,
    threshold,
    verticalTicks,
    verticalLabel,
    horizontalDayTicks,
    horizontalMonthYearTicks,
    showGridLines,
    showGraphDetails = true,
    yDirection = -1,
    toolTip,
    showToolTip,
    hideToolTip,
    graphYMax,
    flexibleHeight,
    isLoading,
  }) => (
    <React.Fragment>
      <ResizeGraphWrapper
        render={layout => {
          const height = layout.height;
          const width = layout.width;

          const filteredVerticalTicks = filterVerticalTicksForViewAndVisibilityTolerance(
            verticalTicks,
            flexibleHeight,
            threshold,
            targetRange,
            Y_AXIS_TICK_VISIBILITY_TOLERANCE,
            graphYMax,
          );

          return (
            <TrendPlotWrapperDivWithLoader
              hasError={isEmpty(measurements) && !isLoading}
              isLoading={isLoading}
            >
              <RenderIf validate={width && height}>
                <Graph
                  viewportRight={width}
                  viewportBottom={height}
                  height={height}
                  anchor="xMidYMid"
                >
                  <TrendAxesPlot
                    verticalTicks={filteredVerticalTicks}
                    verticalLabel={verticalLabel}
                    horizontalDayTicks={horizontalDayTicks}
                    horizontalMonthYearTicks={horizontalMonthYearTicks}
                    width={width * 0.95}
                    height={height}
                    yDirection={yDirection}
                  />
                  <TrendGraphPlot
                    height={height}
                    width={width * X_AXIS_WIDTH}
                    yDirection={yDirection}
                    targetRange={targetRange}
                    threshold={threshold.value}
                    points={graphData}
                    horizontalDayTicks={horizontalDayTicks}
                    showGridLines={showGridLines}
                    onCandleStickMouseOver={showToolTip}
                    onCandleStickMouseOut={hideToolTip}
                    graphYMax={graphYMax}
                  />
                </Graph>
              </RenderIf>
            </TrendPlotWrapperDivWithLoader>
          );
        }}
      />
      <RenderIf
        validate={showGraphDetails && !isEmpty(measurements) && !isLoading}
      >
        <GraphDetailContainer>
          <GraphDetail graphStatistics={graphDetails} graphType="trend" />
        </GraphDetailContainer>
      </RenderIf>
      <RenderIf validate={toolTip.x && toolTip.y}>
        {renderToolTip(toolTip, threshold, targetRange)}
      </RenderIf>
    </React.Fragment>
  ),
);

const renderToolTip = (
  { x, y, data },
  { data: thresholdData },
  { data: targetRangeData },
) => {
  const { stdDev, max, min, mean, count } = data;

  return (
    <ToolTip x={x} y={y}>
      <TrendToolTip
        max={max}
        min={min}
        mean={mean && mean.toFixed(1)}
        numMeasurements={count}
        stDev={
          !stdDev || stdDev === EMPTY_VALUE_PLACEHOLDER
            ? stdDev
            : stdDev.toFixed(1)
        }
        meanColor={getToolTipValueColor(
          mean,
          thresholdData.value,
          targetRangeData,
        )}
        maxColor={getToolTipValueColor(
          max,
          thresholdData.value,
          targetRangeData,
        )}
        minColor={getToolTipValueColor(
          min,
          thresholdData.value,
          targetRangeData,
        )}
      />
    </ToolTip>
  );
};
