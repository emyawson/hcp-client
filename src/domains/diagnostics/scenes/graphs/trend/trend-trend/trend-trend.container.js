import { compose } from 'recompose';
import { connect } from 'react-redux';

import { TrendTrend } from './trend-trend.component';
import { trendTrendConnector } from './trend-trend.selector';

export const TrendTrendContainer = compose(
  connect(
    trendTrendConnector,
    null,
  ),
)(TrendTrend);
