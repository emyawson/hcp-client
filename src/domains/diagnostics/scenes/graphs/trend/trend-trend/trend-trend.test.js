import React from 'react';
import { shallow } from 'enzyme';
import { take } from 'ramda';

import {
  makeOverwrite,
  convertISOToJsGMT,
} from 'src/domains/diagnostics/utils';
import { EMPTY_VALUE_PLACEHOLDER } from 'src/domains/diagnostics/store/constants';

import { selectGraphData, trendTrendConnector } from './trend-trend.selector';
import { TrendTrend } from './trend-trend.component';

const mockedState = {
  stripDelivery: {
    thresholds: {
      actualHyper: {
        preIdealInterval: 100,
        postIdealInterval: 99,
        noctIdealInterval: 99,
      },
      hyper: {
        preIdealInterval: 125,
        postIdealInterval: 99,
        noctIdealInterval: 99,
      },
      hypo: {
        preIdealInterval: 60,
        postIdealInterval: 99,
        noctIdealInterval: 99,
      },
      warning: {
        preIdealInterval: 80,
        postIdealInterval: 99,
        noctIdealInterval: 99,
      },
    },
  },
  ui: {
    patientDashboard: {
      bloodGlucoseUnit: 'mg/dL',
      glucoseMeasurements: [
        {
          value: 212,
          date: convertISOToJsGMT('Wed Nov 08 2017 13:39:25 GMT-0500 (EST)'),
        },
        {
          value: 130,
          date: convertISOToJsGMT('Wed Nov 09 2017 13:39:25 GMT-0500 (EST)'),
        },
        {
          value: 140,
          date: convertISOToJsGMT('Wed Nov 09 2017 13:39:25 GMT-0500 (EST)'),
        },
      ],
      graphStartTime: '0:00',
      showBloodGlucoseLines: true,
      showBloodGlucosePoints: true,
      showBloodGlucoseBeforeMealPoints: true,
      showBloodGlucoseAfterMealPoints: true,
      showCarbohydrates: true,
      showMeanBloodGlucose: true,
    },
    patientDateRange: {
      startDate: convertISOToJsGMT('Mon Nov 06 2017 15:55:53 GMT+0000'),
      endDate: convertISOToJsGMT('Thu Nov 09 2017 23:59:33 GMT+0000'),
    },
  },
};

const mockProps = {
  graphData: [
    {
      data: {
        count: 1,
        max: 212,
        mean: 212,
        min: 212,
        stdDev: EMPTY_VALUE_PLACEHOLDER,
      },
      deviation: 0,
      max: 0.53,
      min: 0.53,
      x: 0.25,
      y: 0.53,
    },
    {
      data: {
        count: 2,
        max: 140,
        mean: 135,
        min: 130,
        stdDev: 7.0710678118654755,
      },
      deviation: 0.017677669529663688,
      max: 0.35,
      min: 0.325,
      x: 0.75,
      y: 0.3375,
    },
  ],
};

const overwriteMockedState = makeOverwrite(mockedState);

describe('Trend Trend Component', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<TrendTrend {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Trend Trend Selector', () => {
  it('should normalize graph data', () => {
    const glucoseMeasurements = [
      {
        value: 212,
        date: convertISOToJsGMT('2017-07-01 12:00:00 GMT+0000'),
      },
      {
        value: 130,
        date: convertISOToJsGMT('2017-07-02 12:00:00 GMT+0000'),
      },
      {
        value: 140,
        date: convertISOToJsGMT('2017-07-02 12:00:00 GMT+0000'),
      },
    ];
    const endDate = convertISOToJsGMT('2017-07-02 23:59:00 GMT+0000');
    const startDate = convertISOToJsGMT('2017-07-01 00:00:00 GMT+0000');
    const { graphData } = mockProps;

    expect(
      selectGraphData.resultFunc(glucoseMeasurements, startDate, endDate),
    ).toEqual(graphData);
  });

  describe('dynamic vertical ticks', () => {
    describe('when the highest measurements value is less than 400', () => {
      const glucoseMeasurements = [
        {
          date: new Date('Wed Nov 08 2017 13:39:25 GMT-0500 (EST)'),
          beforeMeal: true,
          afterMeal: false,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 230,
        },
        {
          date: new Date('Thu Nov 08 2017 15:39:25 GMT-0500 (EST)'),
          beforeMeal: false,
          afterMeal: true,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 240,
        },
        {
          date: new Date('Thu Nov 09 2017 13:39:25 GMT-0500 (EST)'),
          beforeMeal: false,
          afterMeal: true,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 280,
        },
      ];

      it('vertical ticks start from highest value rounded up to a dividend of 50', () => {
        const expected = [
          { gridLine: true, label: '0', value: 0 },
          { gridLine: true, label: '50', value: 50 / 300 },
          { gridLine: true, label: '100', value: 100 / 300 },
          { gridLine: true, label: '150', value: 150 / 300 },
          { gridLine: true, label: '200', value: 200 / 300 },
          { gridLine: true, label: '250', value: 250 / 300 },
          { gridLine: true, label: '300', value: 1 },
        ];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = take(7)(trendTrendConnector(mockState).verticalTicks);
        expect(actual).toEqual(expected);
      });

      it('graph max Y is the highest value rounded up to a dividend of 50', () => {
        const expected = 300;

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = trendTrendConnector(mockState).graphYMax;
        expect(actual).toEqual(expected);
      });

      it('graph data should reflect the dynamic ceiling', () => {
        const day8MeanY = (240 + 230) / 2 / 300;
        const day9MeanY = 280 / 300;
        const expected = [day8MeanY, day9MeanY];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = trendTrendConnector(mockState).graphData.map(x => x.y);
        expect(actual).toEqual(expected);
      });

      it('target range should reflect the dynamic ceiling', () => {
        const warningY = 80 / 300;
        const hyperY = 125 / 300;
        const expected = { min: warningY, max: hyperY };

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = trendTrendConnector(mockState).targetRange;
        expect(actual).toEqual(expect.objectContaining(expected));
      });

      it('threshold should reflect the dynamic ceiling', () => {
        const expected = 60 / 300;

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = trendTrendConnector(mockState).threshold.value;
        expect(actual).toEqual(expected);
      });
    });

    describe('when the highest measurements value is less than the upper limit', () => {
      const glucoseMeasurements = [
        {
          date: new Date('Wed Nov 08 2017 13:39:25 GMT-0500 (EST)'),
          beforeMeal: true,
          afterMeal: false,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 20,
        },
        {
          date: new Date('Wed Nov 08 2017 15:39:25 GMT-0500 (EST)'),
          beforeMeal: true,
          afterMeal: false,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 24,
        },
        {
          date: new Date('Thu Nov 09 2017 13:39:25 GMT-0500 (EST)'),
          beforeMeal: false,
          afterMeal: true,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 30,
        },
      ];

      it('vertical ticks start from the dividend of 50 above upper limit', () => {
        const expected = [
          { gridLine: true, label: '0', value: 0 },
          { gridLine: true, label: '50', value: 1 / 3 },
          { gridLine: true, label: '100', value: 2 / 3 },
          { gridLine: true, label: '150', value: 3 / 3 },
        ];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = take(4)(trendTrendConnector(mockState).verticalTicks);
        expect(actual).toEqual(expected);
      });

      it('graph max Y is the dividend of 50 above upper limit', () => {
        const expected = 150;

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = trendTrendConnector(mockState).graphYMax;
        expect(actual).toEqual(expected);
      });

      it('graph data should reflect the dynamic ceiling', () => {
        const day8MeanY = (20 + 24) / 2 / 150;
        const day9MeanY = 30 / 150;
        const expected = [day8MeanY, day9MeanY];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = trendTrendConnector(mockState).graphData.map(x => x.y);
        expect(actual).toEqual(expected);
      });

      it('target range should reflect the dynamic ceiling', () => {
        const warningY = 80 / 150;
        const hyperY = 125 / 150;
        const expected = { min: warningY, max: hyperY };

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = trendTrendConnector(mockState).targetRange;
        expect(actual).toEqual(expect.objectContaining(expected));
      });

      it('threshold should reflect the dynamic ceiling', () => {
        const expected = 60 / 150;

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = trendTrendConnector(mockState).threshold.value;
        expect(actual).toEqual(expected);
      });
    });

    describe('when the highest measurements value is more than 400', () => {
      const glucoseMeasurements = [
        {
          date: new Date('Wed Nov 08 2017 13:39:25 GMT-0500 (EST)'),
          beforeMeal: true,
          afterMeal: false,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 230,
        },
        {
          date: new Date('Thu Nov 08 2017 15:39:25 GMT-0500 (EST)'),
          beforeMeal: false,
          afterMeal: true,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 240,
        },
        {
          date: new Date('Thu Nov 09 2017 13:39:25 GMT-0500 (EST)'),
          beforeMeal: false,
          afterMeal: true,
          aboveTargetRange: true,
          belowTargetRange: false,
          hypoSymptoms: false,
          value: 420,
        },
      ];

      it('vertical ticks start from 400', () => {
        const expected = [
          { gridLine: true, label: '0', value: 0 },
          { gridLine: true, label: '50', value: 50 / 400 },
          { gridLine: true, label: '100', value: 100 / 400 },
          { gridLine: true, label: '150', value: 150 / 400 },
          { gridLine: true, label: '200', value: 200 / 400 },
          { gridLine: true, label: '250', value: 250 / 400 },
          { gridLine: true, label: '300', value: 300 / 400 },
          { gridLine: true, label: '350', value: 350 / 400 },
          { gridLine: true, label: '400', value: 1 },
        ];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = take(9)(trendTrendConnector(mockState).verticalTicks);
        expect(actual).toEqual(expected);
      });

      it('graph max Y is 400', () => {
        const expected = 400;

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = trendTrendConnector(mockState).graphYMax;
        expect(actual).toEqual(expected);
      });

      it('graph data should reflect the dynamic ceiling', () => {
        const day8MeanY = (240 + 230) / 2 / 400;
        const day9MeanY = 420 / 400;
        const expected = [day8MeanY, day9MeanY];

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });

        const actual = trendTrendConnector(mockState).graphData.map(x => x.y);
        expect(actual).toEqual(expected);
      });

      it('target range should reflect the dynamic ceiling', () => {
        const warningY = 80 / 400;
        const hyperY = 125 / 400;
        const expected = { min: warningY, max: hyperY };

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = trendTrendConnector(mockState).targetRange;
        expect(actual).toEqual(expect.objectContaining(expected));
      });

      it('threshold should reflect the dynamic ceiling', () => {
        const expected = 60 / 400;

        const mockState = overwriteMockedState({
          ui: { patientDashboard: { glucoseMeasurements } },
        });
        const actual = trendTrendConnector(mockState).threshold.value;
        expect(actual).toEqual(expected);
      });
    });
  });
});
