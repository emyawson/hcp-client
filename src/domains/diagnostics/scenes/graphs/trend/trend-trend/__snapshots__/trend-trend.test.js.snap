// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trend Trend Component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withState(withHandlers(Component))
    graphData={
      Array [
        Object {
          "data": Object {
            "count": 1,
            "max": 212,
            "mean": 212,
            "min": 212,
            "stdDev": "-",
          },
          "deviation": 0,
          "max": 0.53,
          "min": 0.53,
          "x": 0.25,
          "y": 0.53,
        },
        Object {
          "data": Object {
            "count": 2,
            "max": 140,
            "mean": 135,
            "min": 130,
            "stdDev": 7.0710678118654755,
          },
          "deviation": 0.017677669529663688,
          "max": 0.35,
          "min": 0.325,
          "x": 0.75,
          "y": 0.3375,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "class",
    "props": Object {
      "graphData": Array [
        Object {
          "data": Object {
            "count": 1,
            "max": 212,
            "mean": 212,
            "min": 212,
            "stdDev": "-",
          },
          "deviation": 0,
          "max": 0.53,
          "min": 0.53,
          "x": 0.25,
          "y": 0.53,
        },
        Object {
          "data": Object {
            "count": 2,
            "max": 140,
            "mean": 135,
            "min": 130,
            "stdDev": 7.0710678118654755,
          },
          "deviation": 0.017677669529663688,
          "max": 0.35,
          "min": 0.325,
          "x": 0.75,
          "y": 0.3375,
        },
      ],
      "toggleToolTip": [Function],
      "toolTip": Object {
        "data": Object {},
        "x": null,
        "y": null,
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "graphData": Array [
          Object {
            "data": Object {
              "count": 1,
              "max": 212,
              "mean": 212,
              "min": 212,
              "stdDev": "-",
            },
            "deviation": 0,
            "max": 0.53,
            "min": 0.53,
            "x": 0.25,
            "y": 0.53,
          },
          Object {
            "data": Object {
              "count": 2,
              "max": 140,
              "mean": 135,
              "min": 130,
              "stdDev": 7.0710678118654755,
            },
            "deviation": 0.017677669529663688,
            "max": 0.35,
            "min": 0.325,
            "x": 0.75,
            "y": 0.3375,
          },
        ],
        "toggleToolTip": [Function],
        "toolTip": Object {
          "data": Object {},
          "x": null,
          "y": null,
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
